### PR TITLE
`Integrade code lifecycle`: Change priority of test exams and practice exercises

### PR DESCRIPTION
**Pull request by BBesrour**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/development-process.html#naming-conventions-for-github-pull-requests).

#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).

### Motivation and Context

Changed priority depending on the programming exercise.
- Real exam: priority 1
- Test exam: priority 2
- Graded course exercise (i.e. before due date): priority 2
- Practice exercise or non-running exams (i.e. after due date): priority 3

### Steps for Testing

You can thoroughly test this locally, follow the steps below, otherwise you can just make sure that everything works on TS3: Use [Test Course Mohamed Bilel Besrour](https://artemis-test3.artemis.cit.tum.de/courses/48). Exams and exercises will be prepared.

Easy testing:
- Go to TS3, use one of the student test users (artemis_test_user_1-5)
- Go to course [Test Course Mohamed Bilel Besrour](https://artemis-test3.artemis.cit.tum.de/courses/48)
- Make a submission for a test exam, real exam, exercise with past due data, exercise with future due date. The exams and exercises should already be there
- Check that the submission worked and you get a result for your submission

Thorough testing (locally):
Prerequisites:
- 1 Admin
- 1 Programming Exercise with due date already passed
- 1 Programming Exercise with due data in the future
- 1 TestExam with programming exercise
- 1 Real Exam with programming exercise

1. Start artemis with these profiles: `dev,localci,localvc,artemis,scheduling,core,ldap-only,local` (we removed buildagent so that buildjobs stay in the queue and are not processed)
2. Make a submission foreach of the programming exercises stated in the prerequisites 
3. Navigate to `Build Overview`
4. Under `Queued Build Jobs`, Make sure that each submission has the correct priority (as stated in the description)

[](https://artemis-test1.artemis.cit.tum.de)
[](https://artemis-test2.artemis.cit.tum.de)
[](https://artemis-test3.artemis.cit.tum.de)
[](https://artemis-test4.artemis.cit.tum.de)
[](https://artemis-test5.artemis.cit.tum.de)
[](https://artemis-test6.artemis.cit.tum.de)
[](https://artemis-test9.artemis.cit.tum.de)

#### Performance Review
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2

## Summary by CodeRabbit

- **New Features**
  - Improved priority calculation for programming exercises in Continuous Integration (CI) management. Exam exercises now have the highest priority, and exercises with past due dates have the lowest priority.

- **Bug Fixes**
  - Enhanced the functionality of test cases by introducing checks and mechanisms for setting build priorities based on specific conditions related to programming exercises and exams.

- **Documentation**
  - Updated system design documentation to reflect changes in priority determination logic for CI build jobs.

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCITriggerService.java [hunk 1]
```diff
@@ -273,4 +273,17 @@ private BuildConfig getBuildConfig(ProgrammingExerciseParticipation participatio
         return new BuildConfig(buildScript, dockerImage, commitHashToBuild, assignmentCommitHash, testCommitHash, branch, programmingLanguage, projectType,
                 staticCodeAnalysisEnabled, sequentialTestRunsEnabled, testwiseCoverageEnabled, resultPaths);
     }
+
+    private int determinePriority(ProgrammingExercise programmingExercise) {
+        if (programmingExercise.isExamExercise()) {
+            boolean isTestExam = programmingExercise.getExerciseGroup().getExam().isTestExam();
+            return isTestExam ? 2 : 1;
+        }
+        else if (programmingExercise.getDueDate() != null && programmingExercise.getDueDate().isBefore(ZonedDateTime.now())) {
```
Comment (JohannesStoehr) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCITriggerService.java:282]: This should also take the participation with a potentially later individual due date into account

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]
See inline comment. We might also want to take a few more metrics into account (in a follow up):
Should optional exercises (i.e. not relevant for points in the course) be less important?
Should the builds of template and solution be more important to ensure an easy workflow for instructors?
Should the build all option in an exercise also be less important than any other build?

**Comment by BBesrour** [comment]
@JohannesStoehr Thanks for the review! Adjusted the code so that individual due dates are taken into account. (used the latter date).
And I was already looking into how we could make the prioritization a bit better, thnx for metrics suggestions 👍

File: src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCITriggerService.java [hunk 2]
```diff
@@ -273,4 +273,29 @@ private BuildConfig getBuildConfig(ProgrammingExerciseParticipation participatio
...
+        // If individual due date is set, use the latter of the two due dates
+        if (programmingExercise.getDueDate() != null && participation.getIndividualDueDate() != null) {
+            dueDate = participation.getIndividualDueDate().isAfter(programmingExercise.getDueDate()) ? participation.getIndividualDueDate() : programmingExercise.getDueDate();
+        }
+        else if (participation.getIndividualDueDate() != null) {
+            dueDate = participation.getIndividualDueDate();
+        }
+        else if (programmingExercise.getDueDate() != null) {
+            dueDate = programmingExercise.getDueDate();
+        }
```
Comment (Strohgelaender) [inline src/main/java/de/tum/in/www1/artemis/service/connectors/localci/LocalCITriggerService.java:288]: Please use the `ExerciseDateService` to determine the due date.

**Review by Strohgelaender — CHANGES_REQUESTED** [review]
Can you also add test cases for the correct priorty calculation?

**Review by JohannesStoehr — APPROVED** [review]
Code

**Review by Strohgelaender — DISMISSED** [review]
code

**Review by pzdr7 — DISMISSED** [review]
Tested locally.

- Practice exercise participation -> 3
- Exercise without due date -> 2
- Exercise with due date in the future -> 2
- Test exam -> 2
- Exam -> 1

Works as expected! ✅

**Review by marlon-luca-bu — APPROVED** [review]
Code

**Review by EneaGore — DISMISSED** [review]
Tested on TS3. Works as expected and priorities are shown correctly on the build queue from admin perspective.

**Pull request merged**